File: packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.localauth;

import static android.content.Context.KEYGUARD_SERVICE;

import android.app.Activity;
import android.app.KeyguardManager;
import android.content.Context;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.biometric.BiometricManager;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Lifecycle;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.embedding.engine.plugins.lifecycle.FlutterLifecycleAdapter;
import io.flutter.plugins.localauth.AuthenticationHelper.AuthCompletionHandler;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import kotlin.Result;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;

/**
 * Flutter plugin providing access to local authentication.
 *
 * <p>Instantiate this in an add to app scenario to gracefully handle activity and context changes.
 */
public class LocalAuthPlugin implements FlutterPlugin, ActivityAware, LocalAuthApi {
  private Activity activity;
  private AuthenticationHelper authHelper;

  @VisibleForTesting final AtomicBoolean authInProgress = new AtomicBoolean(false);

  private Lifecycle lifecycle;
  private BiometricManager biometricManager;
  private KeyguardManager keyguardManager;

  /**
   * Default constructor for LocalAuthPlugin.
   *
   * <p>Use this constructor when adding this plugin to an app with v2 embedding.
   */
  public LocalAuthPlugin() {}

  @Override
  public boolean isDeviceSupported() {
    return isDeviceSecure() || canAuthenticateWithBiometrics();
  }

  @Override
  public boolean deviceCanSupportBiometrics() {
    return hasBiometricHardware();
  }

  @Override
  public @Nullable List<AuthClassification> getEnrolledBiometrics() {
    if (biometricManager == null) {
      return null;
    }
    ArrayList<AuthClassification> biometrics = new ArrayList<>();
    if (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
        == BiometricManager.BIOMETRIC_SUCCESS) {
      biometrics.add(AuthClassification.WEAK);
    }
    if (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)
        == BiometricManager.BIOMETRIC_SUCCESS) {
      biometrics.add(AuthClassification.STRONG);
    }
    return biometrics;
  }

  @Override
  public boolean stopAuthentication() {
    try {
      if (authHelper != null && authInProgress.get()) {
        authHelper.stopAuthentication();
        authHelper = null;
      }
      authInProgress.set(false);
      return true;
    } catch (Exception e) {
      return false;
    }
  }

  @Override
  public void authenticate(
      @NonNull AuthOptions options,
      @NonNull AuthStrings strings,
      @NonNull Function1<? super @NotNull Result<@NotNull AuthResult>, @NotNull Unit> callback) {
    if (authInProgress.get()) {
      ResultUtilsKt.completeWithValue(
          callback, new AuthResult(AuthResultCode.ALREADY_IN_PROGRESS, null));
      return;
    }

    if (activity == null || activity.isFinishing()) {
      ResultUtilsKt.completeWithValue(callback, new AuthResult(AuthResultCode.NO_ACTIVITY, null));
      return;
    }

    if (!(activity instanceof FragmentActivity)) {
      ResultUtilsKt.completeWithValue(
          callback, new AuthResult(AuthResultCode.NOT_FRAGMENT_ACTIVITY, null));
      return;
    }

    if (!isDeviceSupported()) {
      ResultUtilsKt.completeWithValue(
          callback, new AuthResult(AuthResultCode.NO_CREDENTIALS, null));
      return;
    }

    authInProgress.set(true);
    AuthCompletionHandler completionHandler = createAuthCompletionHandler(callback);

    boolean allowCredentials = !options.getBiometricOnly() && canAuthenticateWithDeviceCredential();

    sendAuthenticationRequest(options, strings, allowCredentials, completionHandler);
  }

  @VisibleForTesting
  public @NonNull AuthCompletionHandler createAuthCompletionHandler(
      @NonNull Function1<? super @NotNull Result<@NotNull AuthResult>, @NotNull Unit> callback) {
    return authResult -> onAuthenticationCompleted(callback, authResult);
  }

  @VisibleForTesting
  public void sendAuthenticationRequest(
      @NonNull AuthOptions options,
      @NonNull AuthStrings strings,
      boolean allowCredentials,
      @NonNull AuthCompletionHandler completionHandler) {
    authHelper =
        new AuthenticationHelper(
            lifecycle,
            (FragmentActivity) activity,
            options,
            strings,
            completionHandler,
            allowCredentials);

    authHelper.authenticate();
  }

  void onAuthenticationCompleted(
      @NonNull Function1<? super @NotNull Result<@NotNull AuthResult>, @NotNull Unit> callback,
      AuthResult value) {
    if (authInProgress.compareAndSet(true, false)) {
      ResultUtilsKt.completeWithValue(callback, value);
    }
  }

  @VisibleForTesting
  public boolean isDeviceSecure() {
    if (keyguardManager == null) return false;
    return keyguardManager.isDeviceSecure();
  }

  private boolean canAuthenticateWithBiometrics() {
    if (biometricManager == null) return false;
    return biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
        == BiometricManager.BIOMETRIC_SUCCESS;
  }

  private boolean hasBiometricHardware() {
    if (biometricManager == null) return false;
    return biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
        != BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE;
  }

  @VisibleForTesting
  public boolean canAuthenticateWithDeviceCredential() {
    if (Build.VERSION.SDK_INT < 30) {
      // Checking for device credential only authentication via the BiometricManager
      // is not allowed before API level 30, so we check for presence of PIN, pattern,
      // or password instead.
      return isDeviceSecure();
    }

    if (biometricManager == null) return false;
    return biometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL)
        == BiometricManager.BIOMETRIC_SUCCESS;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    LocalAuthApi.Companion.setUp(binding.getBinaryMessenger(), this);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    LocalAuthApi.Companion.setUp(binding.getBinaryMessenger(), null);
  }

  private void setServicesFromActivity(Activity activity) {
    if (activity == null) return;
    this.activity = activity;
    Context context = activity.getBaseContext();
    biometricManager = BiometricManager.from(activity);
    keyguardManager = (KeyguardManager) context.getSystemService(KEYGUARD_SERVICE);
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    setServicesFromActivity(binding.getActivity());
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(binding);
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    lifecycle = null;
    activity = null;
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    setServicesFromActivity(binding.getActivity());
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(binding);
  }

  @Override
  public void onDetachedFromActivity() {
    lifecycle = null;
    activity = null;
  }

  @VisibleForTesting
  final Activity getActivity() {
    return activity;
  }

  @VisibleForTesting
  void setBiometricManager(BiometricManager biometricManager) {
    this.biometricManager = biometricManager;
  }

  @VisibleForTesting
  void setKeyguardManager(KeyguardManager keyguardManager) {
    this.keyguardManager = keyguardManager;
  }
}

File: packages/local_auth/local_auth_android/android/src/main/kotlin/io/flutter/plugins/localauth/LocalAuthPlugin.kt
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
package io.flutter.plugins.localauth

import android.app.Activity
import android.app.KeyguardManager
import android.content.Context
import android.os.Build
import androidx.biometric.BiometricManager
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.lifecycle.FlutterLifecycleAdapter
import io.flutter.plugins.localauth.LocalAuthApi.Companion.setUp
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

/**
 * Flutter plugin providing access to local authentication.
 *
 * Instantiate this in an add to app scenario to gracefully handle activity and context changes.
 */
class LocalAuthPlugin
/**
 * Default constructor for LocalAuthPlugin.
 *
 * Use this constructor when adding this plugin to an app with v2 embedding.
 */
: FlutterPlugin, ActivityAware, LocalAuthApi {
  internal var activity: Activity? = null
    private set

  private var authHelper: AuthenticationHelper? = null

  internal val authInProgress: AtomicBoolean = AtomicBoolean(false)

  private var lifecycle: Lifecycle? = null
  private var biometricManager: BiometricManager? = null
  private var keyguardManager: KeyguardManager? = null

  override fun isDeviceSupported(): Boolean {
    return this.isDeviceSecure || canAuthenticateWithBiometrics()
  }

  override fun deviceCanSupportBiometrics(): Boolean {
    return hasBiometricHardware()
  }

  override fun getEnrolledBiometrics(): List<AuthClassification>? {
    val manager = biometricManager ?: return null
    return buildList {
      if (manager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK) ==
          BiometricManager.BIOMETRIC_SUCCESS) {
        add(AuthClassification.WEAK)
      }
      if (manager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG) ==
          BiometricManager.BIOMETRIC_SUCCESS) {
        add(AuthClassification.STRONG)
      }
    }
  }

  override fun stopAuthentication(): Boolean {
    try {
      if (authInProgress.get()) {
        authHelper?.stopAuthentication()
        authHelper = null
      }
      return true
    } catch (_: Exception) {
      return false
    }
  }

  override suspend fun authenticate(options: AuthOptions, strings: AuthStrings): AuthResult {
    val currentActivity = activity
    if (authInProgress.get()) {
      return AuthResult(AuthResultCode.ALREADY_IN_PROGRESS, null)
    }

    if (currentActivity?.isFinishing ?: true) {
      return AuthResult(AuthResultCode.NO_ACTIVITY, null)
    }

    if (currentActivity !is FragmentActivity) {
      return AuthResult(AuthResultCode.NOT_FRAGMENT_ACTIVITY, null)
    }

    if (!isDeviceSupported()) {
      return AuthResult(AuthResultCode.NO_CREDENTIALS, null)
    }

    authInProgress.set(true)
    val allowCredentials = !options.biometricOnly && canAuthenticateWithDeviceCredential()

    return suspendCoroutine { continuation ->
      val completionHandler = createAuthCompletionHandler(continuation)
      sendAuthenticationRequest(
          options, strings, allowCredentials, currentActivity, completionHandler)
    }
  }

  internal fun createAuthCompletionHandler(
      continuation: Continuation<AuthResult>
  ): (AuthResult) -> Unit {
    return { authResult ->
      if (authInProgress.compareAndSet(true, false)) {
        continuation.resume(authResult)
      }
    }
  }

  internal fun sendAuthenticationRequest(
      options: AuthOptions,
      strings: AuthStrings,
      allowCredentials: Boolean,
      fragmentActivity: FragmentActivity,
      completionHandler: (AuthResult) -> Unit
  ) {
    val helper =
        AuthenticationHelper(
            lifecycle, fragmentActivity, options, strings, completionHandler, allowCredentials)
    authHelper = helper
    helper.authenticate()
  }

  internal val isDeviceSecure: Boolean
    get() {
      return keyguardManager?.isDeviceSecure ?: false
    }

  private fun canAuthenticateWithBiometrics(): Boolean {
    return biometricManager?.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK) ==
        BiometricManager.BIOMETRIC_SUCCESS
  }

  private fun hasBiometricHardware(): Boolean {
    val manager = biometricManager ?: return false
    return manager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK) !=
        BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE
  }

  internal fun canAuthenticateWithDeviceCredential(): Boolean {
    if (Build.VERSION.SDK_INT < 30) {
      // Checking for device credential only authentication via the BiometricManager
      // is not allowed before API level 30, so we check for presence of PIN, pattern,
      // or password instead.
      return this.isDeviceSecure
    }

    return biometricManager?.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL) ==
        BiometricManager.BIOMETRIC_SUCCESS
  }

  override fun onAttachedToEngine(binding: FlutterPluginBinding) {
    setUp(binding.binaryMessenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
    setUp(binding.binaryMessenger, null)
  }

  private fun setServicesFromActivity(activity: Activity?) {
    if (activity == null) return
    this.activity = activity
    val context = activity.baseContext
    biometricManager = BiometricManager.from(activity)
    keyguardManager = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager?
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    setServicesFromActivity(binding.activity)
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(binding)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    lifecycle = null
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    setServicesFromActivity(binding.activity)
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(binding)
  }

  override fun onDetachedFromActivity() {
    lifecycle = null
    activity = null
  }

  internal fun setBiometricManager(biometricManager: BiometricManager?) {
    this.biometricManager = biometricManager
  }

  internal fun setKeyguardManager(keyguardManager: KeyguardManager?) {
    this.keyguardManager = keyguardManager
  }
}
